File: apex_dashboard/apex_dashboard/page/inventory_dashboard/inventory_dashboard.js
frappe.pages['inventory_dashboard'].on_page_load = function (wrapper) {
    new InventoryDashboard(wrapper);
};

class InventoryDashboard {
    constructor(wrapper) {
        this.wrapper = $(wrapper);

        this.page = frappe.ui.make_app_page({
            parent: wrapper,
            title: 'Inventory Dashboard',
            single_column: true
        });

        const TEMPLATE = `
		<div class="inventory-dashboard dashboard-template">
			<div class="dashboard-header">
				<div class="header-content">
					<button class="btn-glass" id="back-btn">
						<i class="fa fa-arrow-left"></i> Hub
					</button>
					<div class="header-text">
						<h1>Inventory Overview</h1>
						<p class="subtitle">Real-time Stock Value & Alerts</p>
					</div>
				</div>
				<div class="header-actions">
					<div class="total-card" style="padding: 10px 20px; min-width: 200px;">
						<span class="label" style="font-size: 12px; color: #9ca3af; display: block;">Total Stock Value</span>
						<span class="value" id="header-total-value" style="font-size: 20px; font-weight: bold;">Loading...</span>
					</div>
					<button class="btn-glass" id="refresh-btn">
						<i class="fa fa-refresh"></i>
					</button>
					<button class="btn-glass" id="config-btn">
						<i class="fa fa-cog"></i>
					</button>
				</div>
			</div>

			<div id="dashboard-loader" class="dashboard-loader">
				<div class="spinner"></div>
				<p>Fetching inventory data...</p>
			</div>

			<div id="dashboard-content" style="display: none;">
				<!-- Content will be injected here -->
			</div>
			
			<div class="dashboard-footer">
				<p>Last updated: <span id="last-updated">-</span></p>
			</div>
		</div>
		`;

        // Load HTML Template
        $(TEMPLATE).appendTo(this.page.body);

        // Hide standard page title
        if (this.page.page_header) {
            this.page.page_header.hide();
        } else {
            this.wrapper.find('.page-head').hide();
        }

        this.format_currency = this.format_currency.bind(this);
        this.setup_filters();
        this.bind_events();
        this.load_data();
    }

    bind_events() {
        this.wrapper.find('#back-btn').on('click', () => {
            frappe.set_route('apex_dashboards');
        });

        this.wrapper.find('#refresh-btn').on('click', () => {
            this.load_data();
        });

        this.wrapper.find('#config-btn').on('click', () => {
            frappe.set_route('Form', 'Apex Dashboard Config', 'Apex Dashboard Config');
        });
    }

    setup_filters() {
        // Company filter
        this.page.add_field({
            fieldname: 'company',
            label: __('Company'),
            fieldtype: 'Link',
            options: 'Company',
            default: frappe.defaults.get_user_default('Company'),
            change: () => this.load_data()
        });

        // Period filter
        this.page.add_field({
            fieldname: 'period',
            label: __('Period'),
            fieldtype: 'Select',
            options: ['', 'Today', 'This Week', 'This Month', 'Last Month', 'This Year', 'Last Year', 'Fiscal Year', 'Custom', 'All Time'],
            default: 'Today',
            change: () => {
                const period = this.page.fields_dict.period.get_value();

                // Toggle Date Fields
                const isCustom = period === 'Custom';
                if (isCustom) {
                    this.page.fields_dict.from_date.$wrapper.removeClass('hide').show();
                    this.page.fields_dict.to_date.$wrapper.removeClass('hide').show();
                } else {
                    this.page.fields_dict.from_date.$wrapper.addClass('hide').hide();
                    this.page.fields_dict.to_date.$wrapper.addClass('hide').hide();
                }

                // Toggle Fiscal Year Field
                const isFiscalYear = period === 'Fiscal Year';
                if (isFiscalYear) {
                    this.page.fields_dict.fiscal_year.$wrapper.removeClass('hide').show();
                } else {
                    this.page.fields_dict.fiscal_year.$wrapper.addClass('hide').hide();
                    this.page.fields_dict.fiscal_year.set_input(''); // Clear value when hidden
                }

                if (period !== 'Custom' && period !== 'Fiscal Year') {
                    this.load_data();
                }
            }
        });

        // Fiscal Year Filter
        this.page.add_field({
            fieldname: 'fiscal_year',
            label: __('Select Year'),
            fieldtype: 'Link',
            options: 'Fiscal Year',
            change: () => {
                const fiscal_year = this.page.fields_dict.fiscal_year.get_value();
                if (fiscal_year) {
                    this.load_data();
                }
            }
        });

        // Date range filters
        this.page.add_field({
            fieldname: 'from_date',
            label: __('From Date'),
            fieldtype: 'Date',
            change: () => this.load_data()
        });

        this.page.add_field({
            fieldname: 'to_date',
            label: __('To Date'),
            fieldtype: 'Date',
            change: () => this.load_data()
        });

        // Initialize visibility
        this.page.fields_dict.period.$input.trigger('change');
    }

    load_data() {
        const company = this.page.fields_dict.company.get_value();
        const period = this.page.fields_dict.period.get_value();
        const from_date = this.page.fields_dict.from_date.get_value();
        const to_date = this.page.fields_dict.to_date.get_value();
        const fiscal_year = this.page.fields_dict.fiscal_year.get_value();

        this.show_loader(true);

        frappe.call({
            method: 'apex_dashboard.apex_dashboard.page.inventory_dashboard.inventory_dashboard.get_dashboard_data',
            args: {
                company: company,
                period: period,
                from_date: from_date,
                to_date: to_date,
                fiscal_year: fiscal_year
            },
            callback: (r) => {
                this.show_loader(false);
                if (r.message) {
                    this.render(r.message);
                }
            }
        });
    }

    show_loader(show) {
        if (show) {
            this.wrapper.find('#dashboard-loader').show();
            this.wrapper.find('#dashboard-content').hide();
        } else {
            this.wrapper.find('#dashboard-loader').hide();
            this.wrapper.find('#dashboard-content').fadeIn();
        }
    }

    getContrastColor(hexColor) {
        if (!hexColor) return '#ffffff';
        hexColor = hexColor.replace('#', '');
        const r = parseInt(hexColor.substr(0, 2), 16);
        const g = parseInt(hexColor.substr(2, 2), 16);
        const b = parseInt(hexColor.substr(4, 2), 16);
        const luminance = (0.299 * r + 0.587 * g + 0.114 * b) / 255;
        return luminance > 0.5 ? '#1a1a1a' : '#ffffff';
    }

    render(data) {
        const currency = data.currency || 'EGP';
        this.wrapper.find('#last-updated').text(frappe.datetime.now_datetime());
        this.wrapper.find('#header-total-value').text(this.format_currency(data.total_stock.value, currency));

        const content = `
			<div class="inventory-container">
				<!-- Total Stock Card (Top) -->
				<div class="total-stock-card">
					<h2>💰 Total Stock Value</h2>
					<div class="total-value">${this.format_currency(data.total_stock.value, currency)}</div>
					<div class="total-items">${data.total_stock.items} items across all categories</div>
				</div>
				
				<!-- Main Category Cards (Dynamic) -->
				<div class="category-cards">
					${data.groups.map(group => this.render_category_card(group, currency)).join('')}
				</div>
				
				<!-- Alerts Section -->
				${data.alerts.total_alerts > 0 ? this.render_alerts(data.alerts) : ''}
				
				<!-- Bottom Grid -->
				<div class="content-grid">
					<!-- Top Items -->
					<div class="dashboard-card">
						<h3>⭐ Top Items by Value</h3>
						${this.render_top_items(data.top_items, currency)}
					</div>
					
					 <!-- Warehouse Breakdown -->
					<div class="dashboard-card">
						<h3>🏭 Warehouse Breakdown</h3>
						${this.render_warehouses(data.warehouses, currency)}
					</div>
				</div>
			</div>
		`;

        this.wrapper.find('#dashboard-content').html(content);
    }

    render_category_card(group, currency) {
        const textColor = this.getContrastColor(group.color);
        const cardStyle = `
			background: linear-gradient(135deg, ${group.color}40 0%, ${group.color}10 100%);
			border-left: 4px solid ${group.color};
			color: ${textColor};
		`;

        const detailsHtml = this.render_card_details(group.details);

        return `
			<div class="category-card-new" style="${cardStyle}">
				<div class="card-header-new">
					<div class="card-title-group">
						<i class="${group.icon}" style="color: ${group.color}"></i>
						<span>${group.name}</span>
					</div>
					<div class="card-value-new">${this.format_currency(group.value, currency)}</div>
				</div>
				
				<div class="card-stats-new">
					<span>${Math.round(group.qty)} units</span>
					<span>${group.items} types</span>
				</div>
				
				${detailsHtml}
			</div>
		`;
    }

    render_card_details(details) {
        if (!details || details.length === 0) return '';

        const itemsHtml = details.slice(0, 5).map(item => {
            const stockBalanceUrl = `/app/query-report/Stock Balance?item_group=${encodeURIComponent(item.item_group)}`;
            return `
				<div class="card-detail-item clickable" onclick="window.location.href='${stockBalanceUrl}'" title="Click to view Stock Balance Report">
					<span class="detail-name">${item.item_group}</span>
					<div class="detail-stats">
						<span class="detail-qty">${Math.round(item.qty || 0)}</span>
					</div>
				</div>
			`;
        }).join('');

        return `<div class="card-details">${itemsHtml}</div>`;
    }

    render_alerts(alerts) {
        let html = '<div class="alerts-section">';

        if (alerts.out_of_stock && alerts.out_of_stock.length > 0) {
            html += '<div class="alert-card critical">';
            html += '<h3>🔴 Out of Stock / Negative Stock</h3>';
            html += '<div class="alert-list">';
            alerts.out_of_stock.forEach(item => {
                html += `
					<div class="alert-item">
						<span class="item-name">${item.item_name}</span>
						<span class="stock-info">Qty: ${item.actual_qty} ${item.stock_uom}</span>
						<span class="warehouse">${item.warehouse}</span>
					</div>
				`;
            });
            html += '</div></div>';
        }

        if (alerts.low_stock && alerts.low_stock.length > 0) {
            html += '<div class="alert-card warning">';
            html += '<h3>🟡 Low Stock (< 10 units)</h3>';
            html += '<div class="alert-list">';
            alerts.low_stock.forEach(item => {
                html += `
					<div class="alert-item">
						<span class="item-name">${item.item_name}</span>
						<span class="stock-info">${item.actual_qty} ${item.stock_uom}</span>
						<span class="warehouse">${item.warehouse}</span>
					</div>
				`;
            });
            html += '</div></div>';
        }

        html += '</div>';
        return html;
    }

    render_top_items(items, currency) {
        if (!items || items.length === 0) {
            return '<p class="no-data">No items available</p>';
        }

        let html = '<div class="items-list">';
        items.forEach((item, index) => {
            const itemUrl = `/app/query-report/Stock Balance?item_code=${encodeURIComponent(item.item_name)}`;
            html += `
				<div class="item-row clickable" onclick="window.location.href='${itemUrl}'">
					<span class="item-rank">${index + 1}</span>
					<div class="item-info">
						<div class="item-name">${item.item_name}</div>
						<div class="item-group">${item.item_group}</div>
					</div>
					<div class="item-stats">
						<div class="item-qty">${item.actual_qty} ${item.stock_uom}</div>
						<div class="item-value">${this.format_currency(item.stock_value, currency)}</div>
					</div>
				</div>
			`;
        });
        html += '</div>';
        return html;
    }

    render_warehouses(warehouses, currency) {
        if (!warehouses || warehouses.length === 0) {
            return '<p class="no-data">No warehouse data available</p>';
        }

        let html = '<div class="items-list">';
        warehouses.forEach((wh, index) => {
            const warehouseUrl = `/app/query-report/Stock Balance?warehouse=${encodeURIComponent(wh.warehouse)}`;
            html += `
				<div class="item-row clickable" onclick="window.location.href='${warehouseUrl}'">
					<span class="item-rank">${index + 1}</span>
					<div class="item-info">
						<div class="item-name">${wh.warehouse}</div>
					</div>
					<div class="item-stats">
						<div class="item-qty">${wh.total_qty} units</div>
						<div class="item-value">${this.format_currency(wh.total_value, currency)}</div>
					</div>
				</div>
			`;
        });
        html += '</div>';
        return html;
    }

    format_currency(value, currency = 'EGP') {
        if (!value) return `0.00 ${currency}`;
        return `${parseFloat(value).toLocaleString('en-US', {
            minimumFractionDigits: 2,
            maximumFractionDigits: 2
        })} ${currency}`;
    }
}

File: apex_dashboard/apex_dashboard/page/inventory_dashboard/inventory_dashboard.py
import frappe
from frappe import _
from frappe.utils import flt, today
from frappe.query_builder import DocType
from frappe.query_builder.functions import Sum, Count

@frappe.whitelist()
def get_dashboard_data(company=None, period="This Month", from_date=None, to_date=None):
	"""
	Get inventory dashboard data with dynamic card configuration.
	"""
	# Check cache first
	cache_key = f"inventory_{company}_{period}_{from_date}_{to_date}"
	cached_data = frappe.cache().get_value(cache_key)
	if cached_data:
		return cached_data
	
	if not company:
		company = frappe.defaults.get_user_default("Company")

	# Get Company Currency
	currency = frappe.get_value("Company", company, "default_currency") or "EGP"

	# 1. Get Total Stock Value
	Bin = DocType("Bin")
	query_total = (
		frappe.qb.from_(Bin)
		.select(
			Count(Bin.item_code).distinct().as_("items_count"),
			Sum(Bin.stock_value).as_("total_value")
		)
		.where(Bin.actual_qty > 0)
	)
	total_stock = query_total.run(as_dict=True)[0]

	# 2. Define Categories (Groups)
	# We will try to fetch configuration from Apex Dashboard Card if available
	# Otherwise fallback to hardcoded defaults
	
	default_groups = [
		{
			"name": "Machines",
			"title": "🏭 Machines",
			"color": "#E91E63",
			"icon": "fa fa-cogs",
			"item_groups": [
				'3D Resin Printer', 'Metal Printer', 'Dental Milling machines',
				'Zircon Furnace', 'Extra Oral Scanner', 'Intra Oral Scanner'
			]
		},
		{
			"name": "Materials",
			"title": "📦 Materials",
			"color": "#9C27B0",
			"icon": "fa fa-cubes",
			"item_groups": [
				'Zircon Block', 'Zircon Disk', 'PMMA Disk', 'Titanium Disk',
				'Hybird Block', 'Lithium Disilicate Block', 'Lithium Disilicate Ingot',
				'3D Risen Material', 'Implant Accessories', 'Implant Fixture'
			]
		},
		{
			"name": "Software",
			"title": "💻 Software",
			"color": "#2196F3",
			"icon": "fa fa-laptop",
			"item_groups": ['CAD Software', 'CAM Software']
		},
		{
			"name": "Accessories",
			"title": "🔧 Accessories",
			"color": "#FF9800",
			"icon": "fa fa-wrench",
			"item_groups": ['Milling Tools', 'New Spare Parts', 'Suction']
		}
	]

	# Fetch dynamic config if exists
	cards = frappe.get_all("Apex Dashboard Card",
		filters={"category": "Inventory", "is_active": 1},
		fields=["name", "card_title", "color", "icon"]
	)
	
	# Map dynamic cards to groups if they match names, or append new ones?
	# For now, we'll stick to the fixed structure but override visual properties if a card matches the name
	# This is a hybrid approach to keep the logic simple without complex dynamic query generation for now
	
	card_map = {c.name: c for c in cards}
	
	groups_data = []
	
	for group in default_groups:
		# Override with dynamic config if available
		if group["name"] in card_map:
			card = card_map[group["name"]]
			group["title"] = card.card_title
			group["color"] = card.color
			group["icon"] = card.icon
			
		# Fetch data for this group
		group_data = get_group_data(group["item_groups"])
		
		groups_data.append({
			"name": group["title"], # Use title for display
			"color": group["color"],
			"icon": group["icon"],
			"value": group_data["value"],
			"qty": group_data["qty"],
			"items": group_data["items"],
			"details": group_data["details"]
		})

	# 3. Get Alerts (Low Stock & Out of Stock)
	alerts = get_alerts()

	# 4. Get Top Items
	top_items = get_top_items()

	# 5. Get Warehouse Breakdown
	warehouses = get_warehouse_breakdown()

	# Build response data
	data = {
		"currency": currency,
		"total_stock": {
			"items": total_stock.get('items_count') or 0,
			"value": total_stock.get('total_value') or 0
		},
		"groups": groups_data, # Standardized list for dynamic rendering
		"alerts": alerts,
		"warehouses": warehouses,
		"top_items": top_items,
		"period": {
			"from_date": from_date or today(),
			"to_date": to_date or today()
		}
	}
	
	# Cache for 5 minutes
	frappe.cache().set_value(cache_key, data, expires_in_sec=300)
	
	return data

def get_group_data(item_groups):
	"""Helper to fetch stock data for specific item groups using Query Builder"""
	Bin = DocType("Bin")
	Item = DocType("Item")
	
	query = (
		frappe.qb.from_(Bin)
		.join(Item).on(Bin.item_code == Item.name)
		.select(
			Item.item_group,
			Count(Bin.item_code).distinct().as_("count"),
			Sum(Bin.actual_qty).as_("qty"),
			Sum(Bin.stock_value).as_("value")
		)
		.where(Item.item_group.isin(item_groups))
		.where(Bin.actual_qty > 0)
		.where(Item.disabled == 0)
		.where(Bin.warehouse.isin(['Demo Machines - AP', 'Headquarter Warehouse - AP', 'Used Machines - AP']))
		.groupby(Item.item_group)
		.orderby(Sum(Bin.stock_value), order=frappe.qb.desc)
	)
	
	details = query.run(as_dict=True)
	
	return {
		'items': sum(d.get('count', 0) for d in details),
		'qty': sum(d.get('qty', 0) for d in details),
		'value': sum(d.get('value', 0) for d in details),
		'details': details
	}

def get_alerts():
	"""Fetch low stock and out of stock items"""
	Bin = DocType("Bin")
	Item = DocType("Item")
	
	# Low Stock (< 10)
	low_stock = (
		frappe.qb.from_(Bin)
		.join(Item).on(Bin.item_code == Item.name)
		.select(
			Item.item_name,
			Item.item_group,
			Bin.actual_qty,
			Bin.stock_value,
			Bin.warehouse,
			Item.stock_uom
		)
		.where(Bin.actual_qty > 0)
		.where(Bin.actual_qty < 10)
		.where(Item.disabled == 0)
		.orderby(Bin.stock_value, order=frappe.qb.desc)
		.limit(10)
	).run(as_dict=True)
	
	# Out of Stock (<= 0)
	out_of_stock = (
		frappe.qb.from_(Bin)
		.join(Item).on(Bin.item_code == Item.name)
		.select(
			Item.item_name,
			Item.item_group,
			Bin.actual_qty,
			Bin.warehouse,
			Item.stock_uom
		)
		.where(Bin.actual_qty <= 0)
		.where(Item.disabled == 0)
		.orderby(Bin.actual_qty, order=frappe.qb.asc)
		.limit(5)
	).run(as_dict=True)
	
	return {
		"low_stock": low_stock,
		"out_of_stock": out_of_stock,
		"total_alerts": len(low_stock) + len(out_of_stock)
	}

def get_top_items():
	"""Fetch top items by value"""
	Bin = DocType("Bin")
	Item = DocType("Item")
	
	return (
		frappe.qb.from_(Bin)
		.join(Item).on(Bin.item_code == Item.name)
		.select(
			Item.item_name,
			Item.item_group,
			Bin.actual_qty,
			Bin.stock_value,
			Item.stock_uom
		)
		.where(Bin.actual_qty > 0)
		.orderby(Bin.stock_value, order=frappe.qb.desc)
		.limit(10)
	).run(as_dict=True)

def get_warehouse_breakdown():
	"""Fetch warehouse breakdown"""
	Bin = DocType("Bin")
	
	return (
		frappe.qb.from_(Bin)
		.select(
			Bin.warehouse,
			Count(Bin.item_code).distinct().as_("items_count"),
			Sum(Bin.actual_qty).as_("total_qty"),
			Sum(Bin.stock_value).as_("total_value")
		)
		.where(Bin.actual_qty > 0)
		.groupby(Bin.warehouse)
		.orderby(Sum(Bin.stock_value), order=frappe.qb.desc)
		.limit(15)
	).run(as_dict=True)
